### ISSUE TITLE
Idea: augment :- operator semantics to trace source locations

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#2069]
The assign-or-return (`:-`) a.k.a. "elephant" operator is the moral equivalent of exception propagation, but doesn't offer the traceability you tend to get from something like exceptions, which usually have a stack trace attached. We could potentially add some of this value by having the automatically-generated calls to `PropagateFailure` also pass in a description of the source location. Failure-compatible types such as `Result` could then be augmented to track these locations:

```dafny
datatype Result<+T, +R> = Success(value: T) | Failure(error: R, trace: seq<string> := []) {
  function method PropagateFailure<U>(location: string): Result<U, R>
    requires Failure?
  {
    Failure(this.error, [location] + trace)
  }

  // ...
}
```

**Comment by MikaelMayer** [comment]
Perhaps instead of changing result, we could introduce a new datatype? Source code locations are usually private so meant more for developers than end-user consumers of the code. I would see that this datatype would be easily converted to a Result for the end-users, but would be accessible to the developer if they set up specific flags.

```dafny
datatype ResultOrException<+T, +R> = Result(value: T) | Exception(error: R, trace: seq<string> := [])
...
```

**Comment by robin-aws** [comment]
Agreed that it's likely better to have a new datatype for this - I only intended to propose the new functionality in the `PropogateFailure`, with `Result` as an example.

> Source code locations are usually private so meant more for developers than end-user consumers of the code. 

Yes and no, it's pretty standard for language runtimes to attach source locations to errors for debugging purposes, even if those are meaningless to end users of the software. It's pretty critical to have as much detail logged internally as you reasonably can, so you have a hope of debugging production failures.

It's usually the responsibility of the application to not communicate internal exception details when it doesn't make sense - e.g. ideally a web application will say "Oops, something went wrong" rather than rendering the stack trace. :)

**Comment by MikaelMayer** [comment]
Other than my comment, I like the idea !
